portal-bridge/src/census.rs

```diff
@@ -0,0 +1,287 @@
+use alloy_primitives::U256;
+use anyhow::anyhow;
+use delay_map::HashMapDelay;
+use discv5::enr::NodeId;
+use futures::{channel::oneshot, StreamExt};
+use tokio::{
+    sync::mpsc,
+    time::{Duration, Instant},
+};
+use tracing::{error, info};
+
+use ethportal_api::{
+    generate_random_remote_enr,
+    jsonrpsee::http_client::HttpClient,
+    types::{
+        distance::{Distance, Metric, XorMetric},
+        portal::PongInfo,
+    },
+    BeaconContentKey, BeaconNetworkApiClient, Enr, HistoryContentKey, HistoryNetworkApiClient,
+    OverlayContentKey, StateContentKey, StateNetworkApiClient,
+};
+
+/// Ping delay for liveness check of peers in census
+/// Two minutes was chosen somewhat arbitrarily, and can be adjusted
+/// in the future based on performance
+const LIVENESS_CHECK_DELAY: Duration = Duration::from_secs(120);
+
+/// The maximum number of enrs to return in a response,
+/// limiting the number of OFFER requests spawned by the bridge
+/// for each piece of content
+const ENRS_RESPONSE_LIMIT: usize = 8;
+
+/// The census is responsible for maintaining a list of known peers in the network,
+/// checking their liveness, updating their data radius, iterating through their
+/// rfn to find new peers, and providing interested enrs for a given content key.
+pub struct Census {
+    history: Network,
+    state: Network,
+    beacon: Network,
+    census_rx: mpsc::UnboundedReceiver<EnrsRequest>,
+}
+
+impl Census {
+    pub fn new(client: HttpClient, census_rx: mpsc::UnboundedReceiver<EnrsRequest>) -> Self {
+        Self {
+            history: Network::new(client.clone(), Subnetwork::History),
+            state: Network::new(client.clone(), Subnetwork::State),
+            beacon: Network::new(client.clone(), Subnetwork::Beacon),
+            census_rx,
+        }
+    }
+}
+
+impl Census {
+    pub async fn init(&mut self) {
+        // currently, the census is only initialized for the state network
+        info!("Initializing state network census");
+        self.state.init().await;
+    }
+
+    pub async fn run(&mut self) {
+        loop {
+            tokio::select! {
+                // handle enrs request
+                //biased;
+                Some(request) = self.census_rx.recv() => {
+                    let enrs = self.get_interested_enrs(request.content_key).await;
+                    if let Err(err) = request.resp_tx.send(enrs) {
+                        error!("Error sending enrs response: {err:?}");
+                    }
+                }
+                // network is not initialized, this will never yield
```

Comment:
I would remove/change this comment. If we enable history census in the future, it's likely that this comment will not be noticed (and it would be incorrect).

Same for the beacon.